tests/scenarios/delegated-grant.spec.ts

```diff
@@ -0,0 +1,422 @@
+      const alice = await DidKeyResolver.generate();
+      const deviceX = await DidKeyResolver.generate();
+      const bob = await DidKeyResolver.generate();
+
+      // Bob has the chat protocol installed
+      const protocolDefinition = threadRoleProtocolDefinition;
+      const protocol = threadRoleProtocolDefinition.protocol;
+      const protocolsConfig = await TestDataGenerator.generateProtocolsConfigure({
+        author: bob,
+        protocolDefinition
+      });
+      const protocolsConfigureReply = await dwn.processMessage(bob.did, protocolsConfig.message);
+      expect(protocolsConfigureReply.status.code).to.equal(202);
+
+      // Bob starts a chat thread
+      const threadRecord = await TestDataGenerator.generateRecordsWrite({
+        author       : bob,
+        protocol     : protocolDefinition.protocol,
+        protocolPath : 'thread',
+      });
+      const threadRoleReply = await dwn.processMessage(bob.did, threadRecord.message, threadRecord.dataStream);
+      expect(threadRoleReply.status.code).to.equal(202);
+
+      // Bob adds Alice as a participant in the thread
+      const participantRoleRecord = await TestDataGenerator.generateRecordsWrite({
+        author       : bob,
+        recipient    : alice.did,
+        protocol     : protocolDefinition.protocol,
+        protocolPath : 'thread/participant',
+        contextId    : threadRecord.message.contextId,
+        parentId     : threadRecord.message.recordId,
+        data         : new TextEncoder().encode('Alice is my friend'),
+      });
+      const participantRoleReply = await dwn.processMessage(bob.did, participantRoleRecord.message, participantRoleRecord.dataStream);
+      expect(participantRoleReply.status.code).to.equal(202);
+
+      // Bob writes a chat message in the thread
+      const chatRecord = await TestDataGenerator.generateRecordsWrite({
+        author       : bob,
+        protocol     : protocolDefinition.protocol,
+        protocolPath : 'thread/chat',
+        contextId    : threadRecord.message.contextId,
+        parentId     : threadRecord.message.recordId,
+      });
+      const chatRecordReply = await dwn.processMessage(bob.did, chatRecord.message, chatRecord.dataStream);
+      expect(chatRecordReply.status.code).to.equal(202);
+
+      // Alice creates a delegated grant for device X to act as Alice.
+      const scope: PermissionScope = {
+        interface : DwnInterfaceName.Records,
+        method    : DwnMethodName.Write,
+        protocol
+      };
+
+      const grantToDeviceX = await PermissionsGrant.create({
+        delegated   : true, // this is a delegated grant
+        dateExpires : Time.createOffsetTimestamp({ seconds: 100 }),
+        description : 'Allow device X to write as me in chat protocol',
+        grantedBy   : alice.did,
+        grantedTo   : bob.did,
+        grantedFor  : alice.did,
+        scope       : scope,
+        signer      : Jws.createSigner(alice)
```

Comment:
It seems like the intention is to create a grant allowing deviceX to do `RecordsQuery` as Alice. But this is actually creating a grant allowing Bob to `RecordsWrite` as Alice. Below, the grant is being used to authorize deviceX to query, so if this test is passing as it is currently written, that may be a bug.

Reply:
RecordsQueryHandler doesn't seem to be checking that the granteTo of the `DelegatedPermissionsGrant` matches the author of the `RecordsQuery` nor that the `scope` is `interface: 'Records', method: 'Query'`

Reply:
For `RecordsWrites` with delegated grants, we check that `grantedTo` and `author` match in `RecordsWrite.validateIntegrity`. Maybe we could move/add that check to `authenticate` since we already do some verification of the delegated grant in that function. For checking `scope`, we need to add a new method `RecordsGrantAuthorization.authorizeQuery` and call it from `RecordsQueryHandler`.

Reply:
@diehuxx, thanks for spotting the dumb mistake in tests on line 233!  Yeah, I am aware I haven't checked author, which is why the test (line 395) is not implemented like I've done for `RecordsWrite` (line 293), but since I screwed up the positive test and ended up testing for a negative case, I'll just add the logic now, so that `RecordsQuery` is on parity with `RecordsWrite`.

Thanks for calling out on scoping, I intend to consider after all main Records methods are addressed, right now any delegated grant from Alice gives you full power to act as Alice. This is why scoping tests (line 401 ~ 411) are left as `xit()` placeholders.

On moving some checks into `authenticate`, definitely interested in refactoring to reduce code duplication, I'll take a look if it makes sense (or easy) to do this now. In general I expect a refactoring to take place when delegated grant support for `RecordsRead` and `RecordsDelete` are added.

Reply:
> Thanks for calling out on scoping, I intend to consider after all main Records methods are addressed, right now any delegated grant from Alice gives you full power to act as Alice. 

Ahh, I didn't realize that was the intention. Is the intention to continue supporting an all-powerful delegated grant, or is this just an interim state of affairs before we fully set up scoping?

Reply:
Updated the PR. I refactored the delegated grant referential validation into its own method, but not quite ready to move it into `authenticate()` yet, would love to discuss the pros and cons with you offline.

Reply:
> is this just an interim state of affairs before we fully set up scoping?

It can go either way, have yet to discuss if all-powerful grant is needed, but I assume we don't support it for permissions grant, then we don't support when scope is evaluated. But yeah, it speaks to your point about delegated grant being different to permissions grant.

Reply:
Maybe both delegated and regular permissions grants should have an "all-powerful" grant. We can revisit in the future.